onnxruntime/core/framework/kernel_type_str_resolver.h

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <string>
+#include <string_view>
+#include <utility>
+
+#include "gsl/gsl"
+
+#if !defined(ORT_MINIMAL_BUILD)
+#include "onnx/defs/schema.h"
+#endif  // !defined(ORT_MINIMAL_BUILD)
+
+#include "core/common/inlined_containers.h"
+#include "core/common/status.h"
+#include "core/graph/op_identifier.h"
+#include "core/graph/graph.h"
+#include "core/platform/ort_mutex.h"
+
+namespace flatbuffers {
+class FlatBufferBuilder;
+template <typename T>
+struct Offset;
+}  // namespace flatbuffers
+
+namespace onnxruntime {
+
+namespace fbs {
+struct KernelTypeStrResolver;
+}  // namespace fbs
+
+using ArgTypeAndIndex = std::pair<ArgType, size_t>;
+using KernelTypeStrToArgsMap = InlinedHashMap<std::string, InlinedVector<ArgTypeAndIndex>>;
+using OpKernelTypeStrMap = InlinedHashMap<OpIdentifier, KernelTypeStrToArgsMap>;
+
+/**
+ * This class interface provides a way to resolve an op's kernel type string to its associated arguments.
+ *
+ * A 'kernel type string' is a string that is used in kernel def type constraints. In particular, it can be a named
+ * type parameter (such as 'T') specified in the op schema or it can be the name of an input or output parameter.
+ */
+class IKernelTypeStrResolver {
+ public:
+  /**
+   * Resolves an op's kernel type string to its associated arguments.
+   * @param node The op's node.
+   * @param kernel_type_str The op kernel type string.
+   * @param[out] resolved_args The op arguments associated with kernel_type_str.
+   */
+  virtual Status ResolveKernelTypeStr(const Node& node, std::string_view kernel_type_str,
+                                      gsl::span<const ArgTypeAndIndex>& resolved_args) const = 0;
+};
+
+/**
+ * A basic implementation of IKernelTypeStrResolver.
+ *
+ * Supports loading information from op schemas in a full build and saving to/loading from an ORT format model
+ * representation.
+ */
+class KernelTypeStrResolver : public IKernelTypeStrResolver {
+ public:
+  Status ResolveKernelTypeStr(const Node& node, std::string_view kernel_type_str,
+                              gsl::span<const ArgTypeAndIndex>& resolved_args) const override;
+
+#if !defined(ORT_MINIMAL_BUILD)
+
+  /**
+   * Registers kernel type string matching info from an op schema.
+   * This will not overwrite an existing registration for the same op.
+   * @param op_schema The op schema to register.
+   * @param[out] registered Whether the op schema was registered or there was already an existing registration.
+   */
+  Status RegisterOpSchema(const ONNX_NAMESPACE::OpSchema& op_schema, bool* registered = nullptr);
```

Comment:
Isn't this class for Ort format models? If so, does RegsiterOpSchema even belong here?

Reply:
will try to rework the class structure so the intended usage is more obvious. saying one is for ONNX models vs. ORT format models seems reasonable

Reply:
KernelTypeStrResolver is used with ORT format models, both for saving and loading them. when saving (from an ONNX model in a full build), we actually use RegisterOpSchema() on nodes in the graph.
https://github.com/microsoft/onnxruntime/blob/e047701fe4b0372de48d3faa93435a58fbd37ef3/onnxruntime/core/session/inference_session.cc#L601-L609